File: smoke_tests/swift/authorization.swift
// Authorization and Access Control vulnerabilities in Swift
import Foundation

class AuthorizationVulnerabilities {

    var currentUserId: Int = 0
    var isAdmin: Bool = false

    // Test 1: Missing authorization check
    func deleteUser(userId: Int) {
        // VULNERABLE: No authorization check
        UserDatabase.shared.delete(userId: userId)
    }

    // Test 2: Client-side authorization
    func viewAdminPanel() -> String {
        // VULNERABLE: Authorization on client side only
        if UserDefaults.standard.bool(forKey: "isAdmin") {
            return "Admin Panel Content"
        }
        return "Access Denied"
    }

    // Test 3: Insecure Direct Object Reference (IDOR)
    func getUserData(userId: Int) -> UserData? {
        // VULNERABLE: No ownership check
        return UserDatabase.shared.fetch(userId: userId)
    }

    // Test 4: Horizontal privilege escalation
    func updateProfile(userId: Int, data: [String: Any]) {
        // VULNERABLE: User can update any profile
        UserDatabase.shared.update(userId: userId, data: data)
    }

    // Test 5: Vertical privilege escalation
    func setUserRole(userId: Int, role: String) {
        // VULNERABLE: No admin check before role change
        UserDatabase.shared.setRole(userId: userId, role: role)
    }

    // Test 6: Bypassing authorization via parameter manipulation
    func accessResource(resourceId: String, userId: Int) -> Data? {
        // VULNERABLE: userId from request, not session
        return ResourceManager.shared.get(id: resourceId, ownerId: userId)
    }

    // Test 7: Mass assignment with role
    func updateUser(userId: Int, params: [String: Any]) {
        // VULNERABLE: Can set isAdmin via params
        var user = UserDatabase.shared.fetch(userId: userId)
        if let admin = params["isAdmin"] as? Bool {
            user?.isAdmin = admin  // VULNERABLE
        }
        if let role = params["role"] as? String {
            user?.role = role  // VULNERABLE
        }
    }

    // Test 8: File access without authorization
    func downloadFile(filename: String) -> Data? {
        let path = "/uploads/\(filename)"
        // VULNERABLE: No ownership check
        return FileManager.default.contents(atPath: path)
    }

    // Test 9: API endpoint without auth middleware
    func handleApiRequest(endpoint: String, params: [String: Any]) -> [String: Any] {
        // VULNERABLE: No authentication check
        switch endpoint {
        case "/users":
            return ["users": UserDatabase.shared.getAllUsers()]
        case "/admin/settings":
            return ["settings": getSettings()]
        default:
            return ["error": "Not found"]
        }
    }

    // Test 10: Token-based auth bypass
    func validateToken(token: String) -> Bool {
        // VULNERABLE: Accepts empty or special tokens
        if token.isEmpty { return false }
        if token == "debug" { return true }  // BACKDOOR
        return TokenValidator.validate(token)
    }

    // Test 11: Function-level access control missing
    func adminFunction(action: String) {
        // VULNERABLE: No role check
        switch action {
        case "deleteAll":
            UserDatabase.shared.deleteAll()
        case "export":
            exportAllData()
        default:
            break
        }
    }

    // Test 12: Cached authorization
    func checkPermission(userId: Int, resource: String) -> Bool {
        let cacheKey = "\(userId)-\(resource)"
        // VULNERABLE: Cached permissions may be stale
        if let cached = PermissionCache.shared.get(key: cacheKey) {
            return cached
        }
        let result = checkActualPermission(userId: userId, resource: resource)
        PermissionCache.shared.set(key: cacheKey, value: result)
        return result
    }

    private func getSettings() -> [String: Any] { [:] }
    private func exportAllData() {}
    private func checkActualPermission(userId: Int, resource: String) -> Bool { false }
}

struct UserData {
    var id: Int
    var name: String
    var isAdmin: Bool
    var role: String
}

class UserDatabase {
    static let shared = UserDatabase()
    func delete(userId: Int) {}
    func fetch(userId: Int) -> UserData? { nil }
    func update(userId: Int, data: [String: Any]) {}
    func setRole(userId: Int, role: String) {}
    func getAllUsers() -> [UserData] { [] }
    func deleteAll() {}
}

class ResourceManager {
    static let shared = ResourceManager()
    func get(id: String, ownerId: Int) -> Data? { nil }
}

class TokenValidator {
    static func validate(_ token: String) -> Bool { false }
}

class PermissionCache {
    static let shared = PermissionCache()
    func get(key: String) -> Bool? { nil }
    func set(key: String, value: Bool) {}
}

File: smoke_tests/lua/authorization.lua
-- Authorization vulnerabilities in Lua

-- Test 1: Missing authorization check
function delete_user(user_id)
    -- VULNERABLE: No authorization check
    db:execute("DELETE FROM users WHERE id = " .. user_id)
end

-- Test 2: IDOR (Insecure Direct Object Reference)
function get_user_data(user_id)
    -- VULNERABLE: No ownership verification
    return db:query("SELECT * FROM users WHERE id = " .. user_id)
end

-- Test 3: Horizontal privilege escalation
function update_profile(user_id, data)
    -- VULNERABLE: Can update any user's profile
    db:execute("UPDATE users SET name = '" .. data.name .. "' WHERE id = " .. user_id)
end

-- Test 4: Vertical privilege escalation
function set_user_role(user_id, role)
    -- VULNERABLE: No admin check
    db:execute("UPDATE users SET role = '" .. role .. "' WHERE id = " .. user_id)
end

-- Test 5: Client-side authorization
function admin_panel(request)
    -- VULNERABLE: Trusting client-side flag
    if request.params.is_admin == "true" then
        return render_admin_panel()
    end
    return "Access Denied"
end

-- Test 6: Parameter tampering
function access_resource(request)
    -- VULNERABLE: user_id from request, not session
    local user_id = request.params.user_id
    return get_resource_for_user(user_id)
end

-- Test 7: Missing function-level access control
function admin_action(action)
    -- VULNERABLE: No role check
    if action == "delete_all" then
        db:execute("DELETE FROM users")
    elseif action == "export" then
        export_all_data()
    end
end

-- Test 8: Path-based authorization bypass
function access_file(path)
    -- VULNERABLE: No path authorization
    local file = io.open("/uploads/" .. path, "r")
    return file:read("*a")
end

-- Test 9: Broken access control in API
function handle_api_request(endpoint, request)
    -- VULNERABLE: No authorization
    if endpoint == "/users" then
        return get_all_users()
    elseif endpoint == "/admin/settings" then
        return get_admin_settings()
    end
end

-- Test 10: Mass assignment with role
function update_user(user_id, params)
    local user = get_user(user_id)
    -- VULNERABLE: Can set is_admin via params
    for key, value in pairs(params) do
        user[key] = value
    end
    save_user(user)
end

-- Test 11: Debug backdoor
function validate_token(token)
    -- VULNERABLE: Debug backdoor
    if token == "debug" then
        return true
    end
    return verify_jwt(token)
end

-- Test 12: Role check bypass
function require_admin(request)
    local role = request.cookies.role
    -- VULNERABLE: Role from cookie (client-controlled)
    return role == "admin"
end

-- Test 13: ngx.var authorization bypass
function openresty_auth()
    -- VULNERABLE: Trusting request headers for auth
    local user_role = ngx.req.get_headers()["X-User-Role"]
    if user_role == "admin" then
        return true
    end
    return false
end

-- Test 14: Race condition in authorization
function withdraw_funds(user_id, amount)
    local balance = get_balance(user_id)
    -- VULNERABLE: TOCTOU race condition
    if balance >= amount then
        -- Time gap allows race condition
        update_balance(user_id, balance - amount)
        return true
    end
    return false
end

-- Helper functions (stubs)
db = {
    execute = function(query) end,
    query = function(query) return {} end
}

function render_admin_panel() return "Admin Panel" end
function get_resource_for_user(id) return {} end
function export_all_data() end
function get_all_users() return {} end
function get_admin_settings() return {} end
function get_user(id) return {} end
function save_user(user) end
function verify_jwt(token) return false end
function get_balance(id) return 0 end
function update_balance(id, amount) end
